### PR TITLE
feat: finish topic 4 (demo)

### PR DESCRIPTION
**Pull request by hungnguyen9327**

### CODE REVIEW

File: restful_topic4/src/main/resources/application.yaml [hunk 1]
```diff
@@ -0,0 +1,12 @@
+spring:
+  datasource:
+    driver-class-name: org.postgresql.Driver
+    url: jdbc:postgresql://localhost:5432/demo_topic4
+    username: postgres
```
Comment (QuangNguyenVinh-IFD) [inline restful_topic4/src/main/resources/application.yaml:5]: Credentials should not be committed and exposed

File: restful_topic4/src/main/java/com/example/restful_topic4/service/impl/UserServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+import java.util.NoSuchElementException;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.stream.Collectors;
+
+@Service
+public class UserServiceImpl implements UserService<UserDTO> {
+
+    @Autowired
+    private UserRepo repo;
```
Comment (QuangNguyenVinh-IFD) [inline restful_topic4/src/main/java/com/example/restful_topic4/service/impl/UserServiceImpl.java:27]: Should DI by constructor not field

File: restful_topic4/src/main/java/com/example/restful_topic4/model/Role.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
+package com.example.restful_topic4.model;
+
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import jakarta.persistence.*;
```
Comment (QuangNguyenVinh-IFD) [inline restful_topic4/src/main/java/com/example/restful_topic4/model/Role.java:4]: Should not use import wildcard

File: mongodb_topic4/src/main/java/com/example/mongodb_topic4/MongodbTopic4Application.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+import org.springframework.security.crypto.password.PasswordEncoder;
+
+import java.time.OffsetDateTime;
+import java.util.Optional;
+
+@SpringBootApplication
+@EnableMongoAuditing(dateTimeProviderRef = "auditingDateTimeProvider")
+public class MongodbTopic4Application {
+    @Bean
+    public ModelMapper modelMapper() {
```
Comment (QuangNguyenVinh-IFD) [inline mongodb_topic4/src/main/java/com/example/mongodb_topic4/MongodbTopic4Application.java:19]: Should place it to config class

File: mongodb_topic4/src/main/java/com/example/mongodb_topic4/controller/UserController.java [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        this.userService = userService;
+    }
+
+    @GetMapping("/all")
+    public ResponseEntity<List<UserDTO>> getAllUsers() {
+        return new ResponseEntity<List<UserDTO>>(userService.getAll(), HttpStatus.OK);
+    }
+
+    @GetMapping("/all/paging_sorting")
+    public Page<UserDTO> findAllByPage(@RequestParam(defaultValue = "0") int page,
```
Comment (HuyQuangLee) [inline mongodb_topic4/src/main/java/com/example/mongodb_topic4/controller/UserController.java:35]: please add min, max for page

File: mongodb_topic4/src/main/java/com/example/mongodb_topic4/advice/NotFoundAdvice.java [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+
+import com.example.mongodb_topic4.exception.NotFoundException;
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.ControllerAdvice;
+import org.springframework.web.bind.annotation.ExceptionHandler;
+import org.springframework.web.bind.annotation.ResponseBody;
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+@ControllerAdvice
+public class NotFoundAdvice {
```
Comment (HuyQuangLee) [inline mongodb_topic4/src/main/java/com/example/mongodb_topic4/advice/NotFoundAdvice.java:11]: please group with MappingAdvice into 1 class

File: mongodb_topic4/src/main/java/com/example/mongodb_topic4/controller/UserController.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+                                       @RequestParam(defaultValue = "2") int sizePerPage,
+                                       @RequestParam(defaultValue = "ID") SortField sortField,
+                                       @RequestParam(defaultValue = "DESC") Sort.Direction sortDirection) {
+        Pageable pageable = PageRequest.of(page, sizePerPage, sortDirection, sortField.getDatabaseFieldName());
+        return userService.findAllByPage(pageable);
+    }
+
+    @GetMapping("/{id}")
```
Comment (hungnguyen9327) [inline mongodb_topic4/src/main/java/com/example/mongodb_topic4/controller/UserController.java:43]: spcc

File: mongodb_topic4/src/main/java/com/example/mongodb_topic4/model/SortField.java [hunk 8]
```diff
@@ -0,0 +1,15 @@
+package com.example.mongodb_topic4.model;
+
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+
+@Getter
+@AllArgsConstructor
+public enum SortField {
+    ID("id"),
+    USERNAME("name"),
```
Comment (hungnguyen9327) [inline mongodb_topic4/src/main/java/com/example/mongodb_topic4/model/SortField.java:10]: research

**Comment by hungnguyen9327** [comment]
Naming rest api endpoints - Done!

File: mongodb_topic4/src/main/java/com/example/mongodb_topic4/advice/NotFoundAdvice.java (hunk 6 above)
Comment (hungnguyen9327) [reply mongodb_topic4/src/main/java/com/example/mongodb_topic4/advice/NotFoundAdvice.java]: Done !

File: mongodb_topic4/src/main/java/com/example/mongodb_topic4/MongodbTopic4Application.java (hunk 4 above)
Comment (hungnguyen9327) [reply mongodb_topic4/src/main/java/com/example/mongodb_topic4/MongodbTopic4Application.java]: Done!

File: restful_topic4/src/main/java/com/example/restful_topic4/model/Role.java (hunk 3 above)
Comment (hungnguyen9327) [reply restful_topic4/src/main/java/com/example/restful_topic4/model/Role.java]: Done!

File: restful_topic4/src/main/java/com/example/restful_topic4/service/impl/UserServiceImpl.java (hunk 2 above)
Comment (hungnguyen9327) [reply restful_topic4/src/main/java/com/example/restful_topic4/service/impl/UserServiceImpl.java]: Done!

File: restful_topic4/src/main/resources/application.yaml (hunk 1 above)
Comment (hungnguyen9327) [reply restful_topic4/src/main/resources/application.yaml:7]: Done!

File: mongodb_topic4/src/main/java/com/example/mongodb_topic4/controller/UserController.java [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+
+    @GetMapping("/")
+    public Page<UserDTO> getAllByPage(@RequestParam(defaultValue = "0") int page,
+                                      @RequestParam(defaultValue = "2") int sizePerPage,
+                                      @RequestParam(defaultValue = "ID") SortField sortField,
+                                      @RequestParam(defaultValue = "DESC") Sort.Direction sortDirection) {
+        Pageable pageable = PageRequest.of(page, sizePerPage, sortDirection, sortField.getDatabaseFieldName());
+        return userService.findAllByPage(pageable);
+    }
+
```
Comment (hungnguyen9327) [inline mongodb_topic4/src/main/java/com/example/mongodb_topic4/controller/UserController.java:41]: To set default value when we dont enter any information for request params, I can adjust the following statements above as below:

```
@GetMapping("/paging")
public Page<UserDTO> getAllByPage(
    @RequestParam(required = false/*, defaultValue="0"*/) Integer page,
    @RequestParam(required = false/*, defaultValue="2"*/) Integer sizePerPage,
    @RequestParam(required = false/*, defaultValue="ID"*/) SortField sortField,
    @RequestParam(required = false/*, defaultValue="DESC"*/) Sort.Direction sortDirection) {
  Pageable pageable = PageRequest.of(
      page == null ? 0 : (int) page,
      sizePerPage == null ? 2 : (int) sizePerPage,
      sortDirection == null ? Direction.ASC : sortDirection,
      sortField == null ? SortField.ID.getDatabaseFieldName() : sortField.getDatabaseFieldName()
  );
  return userService.findAllByPage(minPage, maxPage, pageable);
}
```

File: mongodb_topic4/src/main/resources/application.yaml [hunk 10]
```diff
@@ -27,4 +27,7 @@ logging:
     com:
       ecommerce: DEBUG
     org:
-      springframework: INFO
\ No newline at end of file
+      springframework: INFO
+
+min-page: 0
+max-page: 3
```
Comment (hungnguyen9327) [inline mongodb_topic4/src/main/resources/application.yaml:33]: Set default value for minimum and maximum page size .

File: mongodb_topic4/src/main/java/com/example/mongodb_topic4/service/impl/UserServiceImpl.java [hunk 11]
```diff
@@ -23,102 +24,109 @@
...
-        return repo.findAll(pageable)
-                .map(user -> modelMapper.map(user, UserDTO.class));
+    Page<UserDTO> pages = repo.findAll(pageable)
+        .map(user -> modelMapper.map(user, UserDTO.class));
+    if (pages.getNumber() > max || pages.getNumber() < min) {
+      throw new ExceedPageBoundException();
+    } else {
+      return pages;
     }
+  }
```
Comment (hungnguyen9327) [inline mongodb_topic4/src/main/java/com/example/mongodb_topic4/service/impl/UserServiceImpl.java]: Integrate with maximum and minimum to limit page that we want to get from pagination.

File: mongodb_topic4/src/main/java/com/example/mongodb_topic4/controller/UserController.java (hunk 5 above)
Comment (HuyQuangLee) [reply mongodb_topic4/src/main/java/com/example/mongodb_topic4/controller/UserController.java]: please handle min,max by validate on param, don't need to code to handle it

Comment (hungnguyen9327) [reply mongodb_topic4/src/main/java/com/example/mongodb_topic4/controller/UserController.java]: My bad, I had just got overthinking. I will add @Size annotation for page parameter, for example: 

`@RequestParam(defaultValue = "0") @Size(min = 0, max = 10) int page,`

**Review by HuyQuangLee — APPROVED** [review]

**Pull request merged**